pkg/nimble/autoadv/include/nimble_autoadv.h

```diff
@@ -0,0 +1,142 @@
+*/
+#ifndef NIMBLE_AUTOADV_START_MANUALLY
+    #define NIMBLE_AUTOADV_START_MANUALLY 0
+#endif
+
+/**
+* @brief    Initialize autoadv module.
+*/
+void nimble_autoadv_init(void);
+
+/**
+ * @brief   Set struct for additional arguments specifying the particulars of
+ *          the advertising procedure. Uses memcpy internally.
+ *
+ *          If there is an active advertising process, it will be restarted.
+ *
+ * @param[in]  params   struct with customized additional arguments
+ */
+void nimble_autoadv_set_ble_gap_adv_params(struct ble_gap_adv_params *params);
+
+/**
+ * @brief   Add a new field to the given advertising data.
+ *
+ *          If there is an active advertising process, it will be restarted.
+ *
+ * @param[in] type      field type to add
+ * @param[in] data      payload for the field
+ * @param[in] data_len  length of the payload in bytes
+ *
+ * @return  BLUETIL_AD_OK if the new field was added
+ * @return  BLUETIL_AD_NOMEM if there is not enough space to write add field
+ */
+int nimble_autoadv_add_field(uint8_t type, const void *data, size_t data_len);
+
+/**
+ * @brief   Set the duration for the advertising procedure.
+ *
+ *          If there is an active advertising process, it will be restarted.
+ *
+ * @param[in]  duration_ms  duration of advertising procedure in ms
+ */
+void nimble_auto_adv_set_adv_duration(int32_t duration_ms);
+
+/**
+ * @brief   Set the callback for gap events. Callback is used for the logic when
+ *          to start the advertising procedure.
+ *
+ *          If there is an active advertising process, it will be restarted.
+ *
+ * @param[in] cb        The callback to associate with this advertising
+ *                      procedure. If advertising ends, the event is reported
+ *                      through this callback. If advertising results in a
+ *                      connection, the connection inherits this callback as its
+ *                      event-reporting mechanism.
+ *
+ * @param[in] cb_arg    The optional argument to pass to the callback function.
+ */
+void nimble_auto_adv_set_gap_cb(ble_gap_event_fn *cb, void *cb_arg);
+
+/**
+ * @brief   Start the automated advertising procedure.
+ *
+ *          Needs to be called manually when NIMBLE_AUTOADV_START_MANUALLY was
+ *          set to true and after every call of nimble_autoadv_stop() to start
+ *          advertising again.
+ */
+void nimble_autoadv_start(void);
+
+/**
+ * @brief   Stop the automated advertising procedure. After calling this, you
+ *          have to call nimble_autoadv_start() manually to restart the process.
+ */
+void nimble_autoadv_stop(void);
+
+/**
+ * @brief   Reset all data regarding the advertising process.
+ *          Following characteristics will be applied:
+ *                  - General discoverable mode (BLE_GAP_DISC_MODE_GEN)
+ *                  - Undirected connectable mode (BLE_GAP_CONN_MODE_UND)
+ *                  - No expiration (BLE_HS_FOREVER)
+ *                  - No name
```

Comment:
Maybe these should be configurable? So having defines like: `NIMBLE_AUTOADV_MODE_DEFAULT`?

Reply:
From these four settings only the name is adjustable right now by setting `NIMBLE_AUTOADV_DEVICE_NAME`. The idea behind this module is to provide a builtin solution for a recurring pattern of advertising a device for an application. 

(1) application starts running -> start advertising
(2) device gets connected -> stop advertising
(3) device gets disconnected -> back to (1) (restart advertising)

So between (3) and (2) advertising shouldn't stop at any time, therefore expiration is fixed to `BLE_HS_FOREVER`.

It should be discoverable (`BLE_GAP_DISC_MODE_GEN` which means `"The general discoverable mode shall be used by devices that need to be discoverable continuously or for no specific condition."`).
Non of the alternatives fits here:
- `BLE_GAP_DISC_MODE_NON` -> not discoverable
- `BLE_GAP_DISC_MODE_LTD` -> discoverable only for short time

All devices should be able to connect to it (`BLE_GAP_CONN_MODE_UND`).
Non of the alternatives fits here:
- `BLE_GAP_CONN_MODE_NON` -> not connectable
- `BLE_GAP_CONN_MODE_DIR` -> used for "fast reconnect" to specific device

Reply:
Ah well I forgot about `nimble_autoadv_set_ble_gap_adv_params`. Using this function you can change the default values. You can find an example here: https://github.com/RIOT-OS/RIOT/pull/13506/commits/db323f6bf22ddb078f2ea530599b04e28e1a5280

Reply:
Ah perfect thanks